A few new shaders

zranger1 opened the pull request:
## New Shaders + debug print() in shader code

#### MatrixScroller
GLSL rewrite of the java pattern, which has been renamed "**OldMatrixScroller**"

**Controls**
- LvlReact - disabled
- FreqReact - disabled
- XPos - x offset 
- YPos - y offset
- Speed - speed
- Angle - rotation angle
- Spin - rotation speed
- Size - grid scale
- Quantity - grid density
- Wow1 - beat reactive "bounce" level
- Wow2 - beat reactive brightness pulse
- WowTrigger - disabled

#### Bubbles
Bubbles of various sizes. (Could we use mothership as a giant bubble ring?)

**Controls**
- LvlReact - disabled
- FreqReact - disabled
- XPos - x offset 
- YPos - y offset
- Speed - speed
- Angle - rotation angle
- Spin - rotation speed
- Size - scale
- Quantity - number of bubbles
- Wow1 - beat reactive bubble bounce level
- Wow2 - background gradient level
- WowTrigger - disabled


#### DotVortex
Many spinning dots. Or arcs, or rings...

**Controls**
- LvlReact - disabled
- FreqReact - disabled
- XPos - x offset 
- YPos - y offset
- Speed - speed
- Angle - rotation angle (3D, around X axis)
- Spin - rotation speed (3D, around X axis)
- Size - scale 
- Quantity - disabled 
- Wow1 - shape (low == rings, high == dots)
- Wow2 - disabled
- WowTrigger - disabled

#### FireEdges
Fire effect for TE's edges. 

**Controls**
- LvlReact - disabled
- FreqReact - disabled
- XPos - disabled
- YPos - disabled
- Speed - speed
- Angle - disabled
- Spin - disabled
- Size - disabled
- Quantity - Amount of fire!
- Wow1 - palette width (higher == more colors)
- Wow2 - mix of palette color vs fire color
- WowTrigger - disabled

#### NeonMess
Swirling translucent ribbons. Very audio reactive.

**Controls**
- LvlReact - audio reactivity
- FreqReact - disabled
- XPos - x offset 
- YPos - y offset
- Speed - speed
- Angle - rotation angle
- Spin - rotation speed
- Size - density field scale
- Quantity - disabled
- Wow1 - contrast/black level
- Wow2 - disabled
- WowTrigger - disabled

#### Debug Print for Shaders
You can now display numeric values from shader code! Once in a while,  this can be hugely helpful. But be warned: Text on the car looks just **TERRIBLE**. Also, printing values that change with every pixel sometimes makes things more confusing. 

To use:
- include the font SDF texture
`#pragma iChannel1 "resources/shaders/textures/fontgenerator.png"`

- include debug print support 
`#include <include/debug.fs>`

- call the print function with your variable, and compose it with your
final color output.  Here's an example from neonmess.fs that displays a
value in cyan.

```
    // Debug print: Show the signal level on the car
    // the print function returns a [0,1] brightness value for the text at the
    // current pixel.
    // Parameters are:
    // print(font texture, screen coords, value, size, left digit count, right digit count)
    //
    float x = print(iChannel1, uv - vec2(0.15, 0.2), signal, 0.075, 2.0, 4.0);

    // multiply text color by brightness level
    vec4 txt = x * vec4(0.0,1.0,1.0,1.0);

    // mix w/final output color by text opacity 
    fragColor = mix(fragColor, txt, txt.a);
```

andrewlook commented on te-app/src/main/java/titanicsend/pattern/jon/FireEdges.java:
```diff
@@ -0,0 +1,101 @@
+  public FireEdges(LX lx) {
+    super(lx, TEShaderView.ALL_POINTS);
+
+    // Controls HOW MUCH FIRE IS FIRE!!!
```
🤣 this comment is awesome :)

andrewlook commented on .run/Titanic's End.run.xml:
```diff
@@ -1,6 +1,7 @@
 <component name="ProjectRunConfigurationManager">
   <configuration default="false" name="Titanic's End" type="Application" factoryName="Application">
     <option name="ALTERNATIVE_JRE_PATH" value="temurin-21" />
+    <option name="ALTERNATIVE_JRE_PATH_ENABLED" value="true" />
```
oops, did you mean to modify the runfile?

zranger1 replied:
No, on Windows, the IDE changes it and it sneaks by me at commit time occasionally.

andrewlook commented on te-app/resources/shaders/bubbles.fs:
```diff
@@ -0,0 +1,224 @@
+    if (d < 0.04) uv = uv + te * d * 2.0;
+
+    // simple antialiasing blur.  Really helps smooth things out on LEDs
+    // just redraws the bubble a few times at small offsets
```
very cool! I'll have to remember this trick

andrewlook commented on te-app/resources/shaders/bubbles.fs:
```diff
@@ -0,0 +1,224 @@
+    // generate several layers of bubbles and calculate their contribution
+    // to the pixel's color. Ideally this would be a loop, but
+    // the parameters are hand tuned for each layer, and GLSL would
+    // just unroll the loop anyway.
```
this is a really nice effect - sometimes I'm making patterns and it feels too "flat" when there are a bunch of objects at roughly the same scale. I love the idea of unrolling a few layers, each with different scale and different noise inputs 🙌

andrewlook commented on te-app/resources/shaders/include/debug.fs:
```diff
@@ -0,0 +1,40 @@
+// To use,
+// include a font SDF texture in your shader's body
+// call print(sampler2D ic, vec2 uv, float v, float s, float l, float r);
+// to print a value (returns intensity as a float)
```
this is SO useful! And also LOL that these are the lengths we have to go to when we want to understand what's happening inside of our shaders 😆

andrewlook reviewed (dismissed):
Wow, incredible work here @zranger1 !! You're seriously raising the bar for what we can do with shaders. Thank you so much for this 🙏

andrewlook commented on te-app/resources/shaders/strangepinwheel.fs:
```diff
@@ -0,0 +1,31 @@
+    float t1 = timebase;
+
+    float d = 1.0 - abs(sin(length(uv) - t1)); // waves
+    float a = abs(mod(atan(uv.y, uv.x) - t1, 2.0 - abs(sin(iTime * 0.2)))); // rays
```
I wonder if it's worth introducing a quantity parameter here?

zranger1 replied:
Oh wow,  I didn't even realize I'd left strangepinwheel in for this PR.    It's a  port of a Pixelblaze pattern I wrote, and needs more inspiration and work before it's show ready for RE, IMO.    I'll take it out, and put it in a later PR when it's more finished.

andrewlook replied:
all good! Hope you don't mind that I saved a copy to experiment with myself a bit too :) I'm realizing that partially-complete shaders are actually pretty creatively motivating for me!

andrewlook commented on te-app/resources/shaders/strangepinwheel.fs:
```diff
@@ -0,0 +1,31 @@
+    float d = 1.0 - abs(sin(length(uv) - t1)); // waves
+    float a = abs(mod(atan(uv.y, uv.x) - t1, 2.0 - abs(sin(iTime * 0.2)))); // rays
+
+    float b = fract(max(a, d));
```
one or two ideas just popped into my head for this one:
- what if we did raw audio waveform along the length of the pinwheel segment, modulated by freqreact?
- and/or, what if wow1/2 added some random noise / wavy deformation so the pinwheel segments curve/twist a bit (like how OutrunGrid has that wavy effect, or how Electric has that spiral domain warping...)

andrewlook commented on te-app/resources/shaders/vortex1.fs:
```diff
@@ -0,0 +1,71 @@
+// we're working w/3d coordinates and will need to handle
```
should we rename this shader to `dot_vortex.fs`, just for the sake of consistency?

andrewlook commented on te-app/resources/shaders/strangepinwheel.fs:
```diff
@@ -0,0 +1,31 @@
+    // normalize, center, correct aspect ratio
+    vec2 uv = fragCoord.xy/iResolution.xy - 0.5;
+    uv.x *= iResolution.x/iResolution.y;
+    uv *= rot(iRotationAngle);
```
I noticed when running this that even when spin=0, it spins counterclockwise for a while, then slows down and reverses and spins clockwise - that surprised me a bit at first, or I thought it was an audio-reactive effect detecting a big change and reversing direction.

I'm guessing that's being driven by this line below: `2.0 - abs(sin(iTime * 0.2))`.

I wonder if maybe keeping the spin parameter on control of the rotation would be more flexible for mapping to stems/etc?

It'd be cool though to maybe have something like `2.0 - abs(sin(iTime * 0.2))` control spacing between the rays (or maybe it could be periodic relative to 4 beats or 16 beats etc..?), the idea being that it'd subtly make the pinwheel segments unequal. Maybe wow2 modulates this, where if it's 0 the pinwheel is perfectly balanced, and if it's higher there's more dynamic ranges between the pinwheel segments?

(Also btw none of these comments I'm making on this shader are merge blockers! This is just me feeling inspired and reacting to what I'm seeing, wanting to get the ideas down while they're fresh in my mind :)

zranger1 commented on te-app/src/main/java/titanicsend/pattern/yoffa/config/ShaderPanelsPatternConfig.java:
```diff
@@ -252,6 +252,53 @@ protected void createShader() {
+      controls.setRange(TEControlTag.WOW1, 0, 0, .09); // beat reactivity
+      controls.setRange(TEControlTag.WOW2, 0.15, 0, 1); // background level
+
+      controls.markUnused(controls.getLXControl(TEControlTag.WOWTRIGGER));
```
A lot of Autoshaders will eventually slow down app load time because to create the class, they have to read and parse the shader, and then read, parse and compile it again when it loads in a channel.   Not a lot of time per shader,  but  en masse, it adds up.   It's easy for me to create the Java classes, so I deliberately leave the autoshaders for other players.

zranger1 replied:
(most of the time, anyway...   neonmess took so long to tune,  I just left it the way it was in case I had to come back to it again.)

andrewlook replied:
ah, that's helpful context, thank you!

zranger1 commented on te-app/resources/shaders/neonmess.fs:
```diff
@@ -0,0 +1,105 @@
+        // that's everywhere on Shadertoy. We repeatedly offset the point by
+        // some function, then reflect and fold the coordinates. In this case, we use
+        // a sin/cos curve for the organic "ribbon" effect, and add in some
+        // audio reactivity.
```
Ray marching is way, way simpler than it sounds.   And you can ray march *anything*, solid or transparent with the same technique.    A lot of the "big" ray marching shaders look complicated b/c they're doing physics based materials or lighting, but the core algorithm is pretty much what you see here.

andrewlook replied:
wow, that's encouraging! I think most of the raymarching stuff I've seen is like Inigo's FBM mountain ranges and forests, which didn't feel quite as relevant to us - thanks for breaking new ground for us here, I think this opens up some really cool possibilities!

zranger1 commented:
@andrewlook,  thanks so much for your review!    I really appreciate your comments, your careful reading and your consistently kind and helpful leadership.   When you get a moment or two, can you re-approve this?    I just renamed "dot_vortex".fs, and deleted strangepinwheel, which really shouldn't have been in the PR to begin with.

andrewlook commented:
> @andrewlook, thanks so much for your review! I really appreciate your comments, your careful reading and your consistently kind and helpful leadership. When you get a moment or two, can you re-approve this? I just renamed "dot_vortex".fs, and deleted strangepinwheel, which really shouldn't have been in the PR to begin with.

Thank you! I'm really enjoying engaging with your PR's :) 

Just re-approved! (Also I changed the settings to not require re-approvals - I've found it slows things down a bit and we all have good etiquette around not getting and approval and then adding a bunch of risky stuff)
